core/src/test/java/kafka/server/share/SharePartitionTest.java

```diff
@@ -3554,6 +3555,907 @@ public void testLsoMovementThenAcquisitionLockTimeoutThenAcknowledgeBatchLastOff
+                OptionalInt.empty(), false));
+        assertTrue(sharePartition.canAcquireRecords());
+
+        sharePartition.acknowledge("member-1", Collections.singletonList(
+                new ShareAcknowledgementBatch(20, 49, Collections.singletonList((byte) 1))));
+
+        assertEquals(RecordState.ACKNOWLEDGED, sharePartition.cachedState().get(40L).offsetState().get(49L).state());
+        assertEquals(RecordState.ACQUIRED, sharePartition.cachedState().get(40L).offsetState().get(50L).state());
+        assertEquals(RecordState.ACQUIRED, sharePartition.cachedState().get(60L).batchState());
+        assertTrue(sharePartition.canAcquireRecords());
+        assertEquals(50, sharePartition.startOffset());
+        assertEquals(79, sharePartition.endOffset());
+        assertEquals(80, sharePartition.nextFetchOffset());
+
+        sharePartition.acquire("member-1", new FetchPartitionData(Errors.NONE, 20, 0, memoryRecords(100, 80),
+                Optional.empty(), OptionalLong.empty(), Optional.empty(),
+                OptionalInt.empty(), false));
+        assertFalse(sharePartition.canAcquireRecords());
+
+        // Final Acknowledgment, all records are acknowledged here.
+        sharePartition.acknowledge("member-1", Collections.singletonList(
+                new ShareAcknowledgementBatch(50, 179, Collections.singletonList((byte) 3))));
+
+        assertEquals(0, sharePartition.cachedState().size());
+        assertTrue(sharePartition.canAcquireRecords());
+        assertEquals(180, sharePartition.startOffset());
+        assertEquals(180, sharePartition.endOffset());
+        assertEquals(180, sharePartition.nextFetchOffset());
+
+        sharePartition.acquire("member-1", new FetchPartitionData(Errors.NONE, 20, 0, memoryRecords(20, 180),
+                Optional.empty(), OptionalLong.empty(), Optional.empty(),
+                OptionalInt.empty(), false));
+
+        assertEquals(1, sharePartition.cachedState().size());
+        assertEquals(RecordState.ACQUIRED, sharePartition.cachedState().get(180L).batchState());
+        assertTrue(sharePartition.canAcquireRecords());
+        assertEquals(180, sharePartition.startOffset());
+        assertEquals(199, sharePartition.endOffset());
+        assertEquals(200, sharePartition.nextFetchOffset());
+    }
+
+    @Test
+    public void testCanAcquireRecordsReturnsTrue() {
+        SharePartition sharePartition = SharePartitionBuilder.builder().build();
+
+        assertEquals(0, sharePartition.startOffset());
+        assertEquals(0, sharePartition.endOffset());
+
+        sharePartition.acquire("member-1", new FetchPartitionData(Errors.NONE, 20, 0, memoryRecords(150, 0),
+                Optional.empty(), OptionalLong.empty(), Optional.empty(),
+                OptionalInt.empty(), false));
+
+        assertTrue(sharePartition.canAcquireRecords());
+        assertEquals(0, sharePartition.startOffset());
+        assertEquals(149, sharePartition.endOffset());
+    }
```

Comment:
How the test case is different from: `testCanAcquireRecordsWithCachedDataAndLimitNotReached`? If we required we can add additional assertions in `testCanAcquireRecordsWithCachedDataAndLimitNotReached` itself.

Reply:
My response is similar to the one [here](https://github.com/apache/kafka/pull/16530#discussion_r1668498496). In one of the tests, I am only testing `canAcquireRecords` and in the other one I test the offset values to see if `maybeUpdateCachedStateAndOffsets` is working correctly.